File: src/pointcloud_to_mesh.py
#!/usr/bin/env python3
import rospy
import numpy as np
from sensor_msgs.msg import PointCloud2
from sensor_msgs import point_cloud2
import open3d as o3d
import os
from moveit_msgs.msg import CollisionObject
from moveit_msgs.srv import ApplyPlanningScene, ApplyPlanningSceneRequest
from shape_msgs.msg import Mesh, MeshTriangle
from geometry_msgs.msg import Point

o3d.utility.set_verbosity_level(o3d.utility.VerbosityLevel.Error)

class PointCloudToMesh:
    def __init__(self):
        rospy.init_node("pointcloud_outlier_removal")
        self.read_mesh = rospy.get_param("~read_mesh", True)
        self.obstacle_subscriber = rospy.Subscriber("/vlm_seg/point_cloud", PointCloud2, self.point_cloud_subscriber)
        self.save_mesh_path = "/ros_ws/src/cartesio_collision_avoidance/meshes/obstacle.stl"
        if not os.path.exists(os.path.dirname(self.save_mesh_path)):
            os.makedirs(os.path.dirname(self.save_mesh_path))
        self.alpha_used = 0.037
        self.eps = 0.1
        self.min_points = 10
        self.ransac_distance_threshold = 0.01
        self.ransac_n = 10
        self.num_iterations = 10
        self.apply_planning_scene_srv = rospy.ServiceProxy("/cartesian/collision_avoidance/apply_planning_scene", ApplyPlanningScene)
        rospy.loginfo("Waiting for service /cartesian/collision_avoidance/apply_planning_scene...")
        self.apply_planning_scene_srv.wait_for_service()
        rospy.loginfo("Service is available.")
        if self.read_mesh:
            rospy.loginfo("Reading mesh from file.")
            self.read_and_publish_mesh()

    def read_and_publish_mesh(self):
        if not os.path.isfile(self.save_mesh_path):
            rospy.logwarn("Mesh file does not exist.")
            return
        o3d_mesh = o3d.io.read_triangle_mesh(self.save_mesh_path)
        if len(o3d_mesh.triangles) == 0 or len(o3d_mesh.vertices) == 0:
            rospy.logwarn("Mesh is empty.")
            return
        rospy.loginfo("Publishing mesh from file.")
        self.publish_obstacle_mesh(o3d_mesh)

    def publish_obstacle_mesh(self, o3d_mesh):
        ros_mesh = self.triangle_mesh_to_shape_msgs(o3d_mesh)
        collision_object = CollisionObject()
        collision_object.id = "obstacle_from_pcd"
        collision_object.header.frame_id = "world"
        collision_object.meshes = [ros_mesh]
        collision_object.operation = CollisionObject.ADD
        req = ApplyPlanningSceneRequest()
        req.scene.is_diff = True
        req.scene.world.collision_objects.append(collision_object)
        try:
            self.apply_planning_scene_srv(req)
            rospy.loginfo("Obstacle mesh applied to the PlanningScene.")
        except rospy.ServiceException as e:
            rospy.logerr(f"Failed to call service: {e}")

    def triangle_mesh_to_shape_msgs(self, o3d_mesh):
        ros_mesh = Mesh()
        vertices = np.asarray(o3d_mesh.vertices)
        triangles = np.asarray(o3d_mesh.triangles)
        for v in vertices:
            p = Point()
            p.x, p.y, p.z = v[0], v[1], v[2]
            ros_mesh.vertices.append(p)
        for t in triangles:
            tri = MeshTriangle()
            tri.vertex_indices[0] = t[0]
            tri.vertex_indices[1] = t[1]
            tri.vertex_indices[2] = t[2]
            ros_mesh.triangles.append(tri)
        return ros_mesh

    def remove_clusters(self, raw_pcd):
        pcd = o3d.geometry.PointCloud(o3d.utility.Vector3dVector(np.array(raw_pcd)))
        pcd.estimate_normals()
        labels = np.array(pcd.cluster_dbscan(eps=self.eps, min_points=self.min_points, print_progress=False))
        if len(labels) == 0 or np.all(labels == -1):
            rospy.logwarn("No clusters found.")
            return None
        valid_labels = labels[labels >= 0]
        if len(valid_labels) == 0:
            rospy.logwarn("No valid clusters found.")
            return None
        largest_cluster_label = np.bincount(valid_labels).argmax()
        largest_cluster_indices = np.where(labels == largest_cluster_label)[0]
        return pcd.select_by_index(largest_cluster_indices)

    def plane_to_mesh(self, pcd):
        plane_model, inliers = pcd.segment_plane(distance_threshold=self.ransac_distance_threshold, ransac_n=self.ransac_n, num_iterations=self.num_iterations)
        pcd_plane = pcd.select_by_index(inliers)
        hull_mesh, _ = pcd_plane.compute_convex_hull()
        hull_mesh.compute_vertex_normals()
        return pcd_plane, hull_mesh

    def save_mesh(self, mesh):
        o3d.io.write_triangle_mesh(self.save_mesh_path, mesh)
        rospy.loginfo("Mesh saved.")

    def reduce_density(self, pcd):
        return pcd.voxel_down_sample(voxel_size=0.05)

    def point_cloud_subscriber(self, msg):
        if self.read_mesh:
            return
        rospy.loginfo("Point cloud received.")
        raw_data = list(point_cloud2.read_points(msg, field_names=("x", "y", "z"), skip_nans=True))
        if not raw_data:
            rospy.logwarn("No points in the point cloud.")
            return
        obstacle = self.remove_clusters(raw_data)
        rospy.loginfo("Clusters filtered.")
        if obstacle is None:
            return
        obstacle = self.reduce_density(obstacle)
        rospy.loginfo("Density reduced.")
        _, mesh = self.plane_to_mesh(obstacle)
        self.publish_obstacle_mesh(mesh)

if __name__ == "__main__":
    PointCloudToMesh()
    rospy.spin()

File: src/pointcloud_to_mesh.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <sensor_msgs/point_cloud2_iterator.h>
#include <moveit_msgs/CollisionObject.h>
#include <moveit_msgs/ApplyPlanningScene.h>
#include <moveit_msgs/ApplyPlanningSceneRequest.h>
#include <moveit_msgs/ApplyPlanningSceneResponse.h>
#include <geometry_msgs/Point.h>
#include <shape_msgs/Mesh.h>
#include <shape_msgs/MeshTriangle.h>
#include <open3d/Open3D.h>
#include <open3d/geometry/PointCloud.h>
#include <open3d/geometry/TriangleMesh.h>
#include <open3d/io/IO.h>
#include <open3d/io/TriangleMeshIO.h>
#include <fstream>
#include <unordered_map>
#include <string>
#include <vector>

class PointCloudToMesh
{
public:
    PointCloudToMesh(ros::NodeHandle& nh)
    {
        ros::NodeHandle pnh("~");
        pnh.param<bool>("read_mesh", read_mesh_, false);
        cloud_sub_ = nh.subscribe("/vlm_seg/point_cloud", 1, &PointCloudToMesh::pointCloudCallback, this);
        save_mesh_path_ = "/ros_ws/src/cartesio_collision_avoidance/meshes/obstacle.stl";
        alpha_used_ = 0.037;
        eps_ = 0.1;
        min_points_ = 10;
        ransac_distance_threshold_ = 0.01;
        ransac_n_ = 10;
        num_iterations_ = 10;
        std::string srv_name = "/cartesian/collision_avoidance/apply_planning_scene";
        ros::service::waitForService(srv_name);
        apply_planning_scene_client_ = nh.serviceClient<moveit_msgs::ApplyPlanningScene>(srv_name);
        if (read_mesh_) readAndPublishMesh();
    }

private:
    void pointCloudCallback(const sensor_msgs::PointCloud2ConstPtr& msg)
    {
        if (read_mesh_) return;
        open3d::geometry::PointCloud cloud = rosMsgToOpen3d(*msg);
        if (cloud.points_.empty()) return;
        auto largest_cluster = removeLargestCluster(cloud);
        if (!largest_cluster || largest_cluster->points_.empty()) return;
        largest_cluster = largest_cluster->VoxelDownSample(0.05);
        auto mesh = planeToMesh(*largest_cluster);
        if (!mesh) return;
        publishObstacleMesh(*mesh);
        open3d::io::WriteTriangleMesh(save_mesh_path_, *mesh, true);
    }

    open3d::geometry::PointCloud rosMsgToOpen3d(const sensor_msgs::PointCloud2& ros_cloud)
    {
        open3d::geometry::PointCloud cloud;
        sensor_msgs::PointCloud2ConstIterator<float> iter_x(ros_cloud, "x");
        sensor_msgs::PointCloud2ConstIterator<float> iter_y(ros_cloud, "y");
        sensor_msgs::PointCloud2ConstIterator<float> iter_z(ros_cloud, "z");
        for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z)
        {
            if (!std::isfinite(*iter_x) || !std::isfinite(*iter_y) || !std::isfinite(*iter_z)) continue;
            cloud.points_.emplace_back(
                Eigen::Vector3d(static_cast<double>(*iter_x),
                                static_cast<double>(*iter_y),
                                static_cast<double>(*iter_z)));
        }
        return cloud;
    }

    std::shared_ptr<open3d::geometry::PointCloud> removeLargestCluster(const open3d::geometry::PointCloud& input_pcd)
    {
        auto pcd_ptr = std::make_shared<open3d::geometry::PointCloud>(input_pcd);
        pcd_ptr->EstimateNormals();
        std::vector<int> labels = pcd_ptr->ClusterDBSCAN(eps_, min_points_, false);
        if (labels.empty()) return nullptr;
        std::unordered_map<int, int> label_counts;
        for (auto lbl : labels) if (lbl >= 0) label_counts[lbl]++;
        if (label_counts.empty()) return nullptr;
        int largest_label = -1;
        int max_count = 0;
        for (auto& kv : label_counts) if (kv.second > max_count) { largest_label = kv.first; max_count = kv.second; }
        std::vector<size_t> indices;
        indices.reserve(max_count);
        for (size_t i = 0; i < labels.size(); ++i) if (labels[i] == largest_label) indices.push_back(i);
        return pcd_ptr->SelectByIndex(indices);
    }

    std::shared_ptr<open3d::geometry::TriangleMesh> planeToMesh(const open3d::geometry::PointCloud& pcd)
    {
        if (pcd.points_.empty()) return nullptr;
        auto pcd_ptr = std::make_shared<open3d::geometry::PointCloud>(pcd);
        Eigen::Vector4d plane_model;
        std::vector<size_t> inliers;
        std::tie(plane_model, inliers) = pcd_ptr->SegmentPlane(ransac_distance_threshold_, ransac_n_, num_iterations_);
        if (inliers.empty()) return nullptr;
        auto plane_pcd = pcd_ptr->SelectByIndex(inliers);
        std::shared_ptr<open3d::geometry::TriangleMesh> hull_mesh;
        std::vector<size_t> hull_indices;
        std::tie(hull_mesh, hull_indices) = plane_pcd->ComputeConvexHull();
        if (!hull_mesh) return nullptr;
        hull_mesh->ComputeVertexNormals();
        return hull_mesh;
    }

    shape_msgs::Mesh triangleMeshToShapeMsg(const open3d::geometry::TriangleMesh& o3d_mesh)
    {
        shape_msgs::Mesh ros_mesh;
        ros_mesh.vertices.resize(o3d_mesh.vertices_.size());
        ros_mesh.triangles.resize(o3d_mesh.triangles_.size());
        for (size_t i = 0; i < o3d_mesh.vertices_.size(); ++i)
        {
            geometry_msgs::Point p;
            p.x = o3d_mesh.vertices_;
            p.y = o3d_mesh.vertices_;
            p.z = o3d_mesh.vertices_;
            ros_mesh.vertices[i] = p;
        }
        for (size_t i = 0; i < o3d_mesh.triangles_.size(); ++i)
        {
            shape_msgs::MeshTriangle tri;
            tri.vertex_indices);
            tri.vertex_indices);
            tri.vertex_indices);
            ros_mesh.triangles[i] = tri;
        }
        return ros_mesh;
    }

    void publishObstacleMesh(const open3d::geometry::TriangleMesh& o3d_mesh)
    {
        moveit_msgs::CollisionObject co;
        co.id = "obstacle_from_pcd";
        co.header.frame_id = "world";
        co.operation = moveit_msgs::CollisionObject::ADD;
        co.meshes.push_back(triangleMeshToShapeMsg(o3d_mesh));
        moveit_msgs::ApplyPlanningScene srv;
        srv.request.scene.is_diff = true;
        srv.request.scene.world.collision_objects.push_back(co);
        apply_planning_scene_client_.call(srv);
    }

    void readAndPublishMesh()
    {
        std::ifstream f(save_mesh_path_.c_str());
        if (!f.good()) return;
        auto o3d_mesh = std::make_shared<open3d::geometry::TriangleMesh>();
        if (!open3d::io::ReadTriangleMesh(save_mesh_path_, *o3d_mesh)) return;
        if (o3d_mesh->triangles_.empty() || o3d_mesh->vertices_.empty()) return;
        publishObstacleMesh(*o3d_mesh);
    }

    ros::Subscriber cloud_sub_;
    ros::ServiceClient apply_planning_scene_client_;
    bool read_mesh_;
    std::string save_mesh_path_;
    double alpha_used_, eps_, ransac_distance_threshold_;
    int min_points_, ransac_n_, num_iterations_;
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "pointcloud_outlier_removal_cpp");
    ros::NodeHandle nh;
    PointCloudToMesh node(nh);
    ros::spin();
    return 0;
}
